aten/src/ATen/cpu/vec256/vec256_qint.h

```diff
@@ -154,6 +154,223 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
+  // ( -, -, -, -, z7 + z3, z6 + z2, z5 + z1, z4 + z0)
+  const __m128i xEightHalfSum = _mm_add_epi16(xEightLo, xEightHi);
+  // ( -, -, -, -, -, -, z5 + z1, z4 + z0)
+  const __m128i xEightDualLo = xEightHalfSum;
+  // ( -, -, -, -, -, -, z7 + z3, z6 + z2)
+  const __m128i xEightDualHi = _mm_shuffle_epi32(xEightHalfSum, _MM_SHUFFLE(1, 1, 1, 1));
+  // ( -, -, -, -, -, -, z5 + z1 + z7 + z3, z4 + z0 + z6 + z2)
+  const __m128i xEightDualSum = _mm_add_epi16(xEightDualLo, xEightDualHi);
+  // ( -, -, -, -, -, -, -, z4 + z0 + z6 + z2)
+  const __m128i xLo = xEightDualSum;
+  // ( -, -, -, -, -, -, -, z5 + z1 + z7 + z3)
+  const __m128i xHi = _mm_shufflelo_epi16(xEightDualSum, _MM_SHUFFLE(1, 1, 1, 1));
+  // sum
+  const __m128i sum = _mm_add_epi16(xLo, xHi);
+  // casting is necessary to handle the negative cases properly
+  int16_t res = _mm_extract_epi16(sum, 0);
+  return static_cast<int64_t>(res);
+}
+
+// horizontal sums unsigned i16
+// x = (x15, ..., x0)
+int64_t custom_mm256_hsum_epu16_overflow(__m256i x) {
+  // (x15, ..., x8), int16
+  const __m128i xHalf1 = _mm256_extractf128_si256(x, 1);
+  // (x7, ..., x0), int16
+  const __m128i xHalf2 = _mm256_castsi256_si128(x);
+  // (x15, ..., x8), int32
+  const __m256i xHalf1_32 = _mm256_cvtepu16_epi32(xHalf1);
+  // (x7, ..., x0), int32
+  const __m256i xHalf2_32 = _mm256_cvtepu16_epi32(xHalf2);
+  // (x15 + x7, ..., x8 + x0) = (y7, ..., y0)
+  const __m256i xSum = _mm256_add_epi32(xHalf1_32, xHalf2_32);
+  // ( -, -, -, -, y3, y2, y1, y0)
+  const __m256i xHalfLo = xSum;
+  // ( -, -, -, -, y7, y6, y5, y4)
+  const __m256i xHalfHi = _mm256_permute2f128_si256(xSum, xSum, _MM_SHUFFLE(0, 1, 0, 1));
+  // ( -, -, -, -, y3 + y7, y2 + y6, y1 + y5, y0 + y4)
+  const __m256i xHalfSum = _mm256_add_epi32(xHalfLo, xHalfHi);
+  // ( -, -, -, -, -, -, y1 + y5, y0 + y4)
+  const __m256i xQuadLo = xHalfSum;
+  // ( -, -, -, -, -, -, y3 + y7, y2 + y6)
+  const __m256i xQuadHi = _mm256_permute4x64_epi64(xHalfSum, _MM_SHUFFLE(1, 1, 1, 1));
+  // ( -, -, -, -, -, -, y1 + y5 + y3 + y7, y0 + y4 + y2 + y6)
+  const __m256i sumQuad = _mm256_add_epi32(xQuadLo, xQuadHi);
+  // ( -, -, -, -, -, -, -, y0 + y4 + y2 + y6)
+  const __m256i lo = sumQuad;
+  // ( -, -, -, -, -, -, -, y1 + y5 + y3 + y7)
+  const __m256i hi = _mm256_shuffle_epi32(sumQuad, _MM_SHUFFLE(1, 1, 1, 1));
+  const __m256i sum = _mm256_add_epi32(lo, hi);
+  uint32_t res = _mm256_extract_epi32(sum, 0);
+  return static_cast<int64_t>(res);
+}
+
+// horizontal sums unsigned i8
+// x = (x31, ..., x0)
+int64_t custom_mm256_hsum_epu8_overflow(__m256i x) {
+  // upsample to int16 to avoid overflow
+  // (x31, ..., x16), int8
+  const __m128i xHalf1 = _mm256_extractf128_si256(x, 1);
+  // (x15, ..., x0), int8
+  const __m128i xHalf2 = _mm256_castsi256_si128(x);
+  // (x31, ... x16), int16
+  const __m256i xHalf1_64 = _mm256_cvtepu8_epi16(xHalf1);
+  // (x15, ..., x0), int16
+  const __m256i xHalf2_64 = _mm256_cvtepu8_epi16(xHalf2);
```

Comment:
Any reason not to use _mm256_hadd_epi16 on xHalf1_64 and xHalf2_64 and then forming a tree of hadds?

Reply:
just to clarify, you mean why not call custom_mm256_hsum_epu16_overflow from here?  If yes - it would be slower (can't remember by how much but I did measure it in my adhoc tester and it was significant) because that functions widens the inputs again, and we only need to widen once to ensure no overflow.

Reply:
I think it was more than 20% slower, but can run again if needed for the exact #

Reply:
I meant the horizontal add intrinsic itself on 16-bit values.

Reply:
ah, I didn't know about `hadd` (was searching for `hsum`).  Thanks for the tip! Yeah, that should improve things, along with your other suggestion - will check it out and benchmark